### PR TITLE
Refactor DataHelper

### PR DESCRIPTION
**Pull request by carlonzo**
My goal here is to remove the useless enum DataType replaced with a char, refactor the addType method (I think is more readable now), and speed up a bit the parse of the String to a DataInfo using the index method with a start position and removing the split. now the String is visited only once until "DELIMITER". Hopefully it makes sense.

### CODE REVIEW

**Comment by carlonzo** [comment]
My goal here is to remove the useless enum DataType replaced with a char, refactor the addType method (I think is more readable now), and speed up a bit the parse of the String to a DataInfo using the index method with a start position and removing the split. now the String is visited only once until "DELIMITER"

File: hawk/src/test/java/com/orhanobut/hawk/DataHelperTest.java [hunk 1]
```diff
@@ -133,7 +133,33 @@
     assertThat(dataInfo.isNewVersion()).isTrue();
     assertThat(dataInfo.getKeyClazz().getName()).isEqualTo(clazz);
     assertThat(dataInfo.getValueClazz()).isNull();
-    assertThat(dataInfo.getDataType()).isEqualTo(DataType.LIST);
+    assertThat(dataInfo.getDataType()).isEqualTo(DataHelper.DATATYPE_LIST);
+  }
+
+  @Test public void getOldDataInfoAsList() {
```
Comment (carlonzo) [inline hawk/src/test/java/com/orhanobut/hawk/DataHelperTest.java]: missing tests for old version of DataInfo

File: hawk/src/test/java/com/orhanobut/hawk/DataHelperTest.java [hunk 2]
```diff
@@ -34,7 +34,7 @@
   }
 
   @Test public void testOldVersionCheck() {
-    DataInfo info = DataHelper.getDataInfo("java.lang.String##00@asdfjasdf");
```
Comment (carlonzo) [inline hawk/src/test/java/com/orhanobut/hawk/DataHelperTest.java]: reading the old implementation of the `addType` method, the String had a different format ( https://github.com/orhanobut/hawk/blob/eaf420196396c341b62f7ced6c86df207cf8f91c/hawk/src/main/java/com/orhanobut/hawk/DataUtil.java#L65 ). Am I missing something?

File: hawk/src/main/java/com/orhanobut/hawk/DataHelper.java [hunk 3]
```diff
@@ -40,57 +37,74 @@ static DataInfo getDataInfo(String storedText) {
...
+    }
+
+    String valueClass = storedText.substring(start, index);
+
+    char dataType = storedText.charAt(++index);
+
+    //skip NEW_VERSION + DELIMITER
+    index += 3;
 
-    String[] infos = text.split(INFO_DELIMITER);
```
Comment (carlonzo) [inline hawk/src/main/java/com/orhanobut/hawk/DataHelper.java]: a split version uses a regex to find matches. it visits the String till the end. removing if would save few cpu cycles.

File: hawk/src/main/java/com/orhanobut/hawk/DataHelper.java [hunk 4]
```diff
@@ -40,57 +37,74 @@ static DataInfo getDataInfo(String storedText) {
     if (TextUtils.isEmpty(storedText)) {
       throw new NullPointerException("Text should not be null or empty");
     }
-    int index = storedText.indexOf(DELIMITER);
+
+    int index = storedText.indexOf(INFO_DELIMITER, 0);
```
Comment (orhanobut) [inline hawk/src/main/java/com/orhanobut/hawk/DataHelper.java]: as default, starting index is 0, could stay like this.

**Comment by orhanobut** [comment]
I need to spend quite a time for this PR.

File: hawk/src/main/java/com/orhanobut/hawk/DataHelper.java (hunk 4 above)
Comment (carlonzo) [inline hawk/src/main/java/com/orhanobut/hawk/DataHelper.java]: Yeah.. :D dont ask me why I wrote that

File: hawk/src/main/java/com/orhanobut/hawk/DataHelper.java (hunk 3 above)
Comment (carlonzo) [inline hawk/src/main/java/com/orhanobut/hawk/DataHelper.java]: short story: moved from O(n) to O(1)

**Comment by carlonzo** [comment]
updates? :)

File: hawk/src/test/java/com/orhanobut/hawk/DataHelperTest.java [hunk 5]
```diff
@@ -27,7 +27,7 @@
   }
 
   @Test public void testNewVersionCheck() {
-    DataInfo info = DataHelper.getDataInfo("java.lang.String##00V@asdfjasdf");
```
Comment (carlonzo) [inline hawk/src/test/java/com/orhanobut/hawk/DataHelperTest.java]: why the `00V` ?

File: hawk/src/test/java/com/orhanobut/hawk/HawkEncoderTest.java [hunk 6]
```diff
@@ -102,7 +102,7 @@
 
   @Test public void decodeObject() throws Exception {
     String clazz = "java.lang.String";
-    String info = "00V";
```
Comment (carlonzo) [inline hawk/src/test/java/com/orhanobut/hawk/HawkEncoderTest.java]: again, why the `00V`? the dataType is only one char.

**Comment by carlonzo** [comment]
Hey @orhanobut I sent this PR in December. is there any chances you can review it?

**Comment by orhanobut** [comment]
Hi @carlonzo I'm really sorry not being able to check the PR. This PR actually do some changes on the an important and I couldn't find a good time with a fresh mind to check it. I'll do a review and raise my concerns, then together we can have a result.

**Comment by carlonzo** [comment]
closing. some of my suggested changes have been already applied on a your commit.

**Pull request closed without merging**